src/lib/markbind/src/preprocessors/variablePreprocessor.js

```diff
@@ -0,0 +1,468 @@
+          + `Missing 'name' or 'id' in variable for ${includeElement.attribs.src} include.`);
+        return;
+      }
+      if (!variables[variableName]) {
+        variables[variableName] = cheerio.html(child.children);
+      }
+    });
+  }
+
+
+  /**
+   * Extract variables from an include element.
+   * Subroutine for {@link renderIncludeFile}
+   * @param includeElement include element to extract variables from
+   * @param parentVariables defined by parent pages, which have priority
+   */
+  static extractIncludeVariables(includeElement, parentVariables) {
+    const includedVariables = { ...parentVariables };
+    VariablePreprocessor.extractIncludeInlineVariables(includeElement, includedVariables);
+    VariablePreprocessor.extractIncludeChildElementVariables(includeElement, includedVariables);
+
+    return includedVariables;
+  }
+
+
+  /*
+   * --------------------------------------------------
+   * Combined rendering methods
+   */
+
+
+  /**
+   * Renders an include file with the supplied context, returning the rendered
+   * content and new context with respect to the child content.
+   * @param filePath of the included file source
+   * @param node of the include element
+   * @param context object containing the variables for the current context
+   * @param asIfAt where the included file should be rendered from
+   */
+  renderIncludeFile(filePath, node, context, asIfAt = filePath) {
+    try {
+      // Extract included variables from the include element, merging with the parent context variables
+      const includeVariables = VariablePreprocessor.extractIncludeVariables(node, context.variables);
+
+      // Extract page variables from the included file
+      const fileContent = fs.readFileSync(filePath, 'utf8');
+      const userDefinedVariables = this.getContentParentSiteVariables(asIfAt);
+      const pageVariables = this.extractPageVariables(asIfAt, fileContent, includeVariables);
+
+      // Render the included content with the extracted variables
+      const content = njUtil.renderRaw(fileContent, {
+        ...pageVariables,
+        ...includeVariables,
+        ...userDefinedVariables,
+      });
+
+      // Set up the next child context
+      const childContext = _.cloneDeep(context);
+      childContext.cwf = asIfAt;
+      childContext.variables = includeVariables;
+
+      return { content, childContext };
+    } catch (e) {
+      const missingReferenceErrorMessage = `Missing reference in: ${node.attribs[ATTRIB_CWF]}`;
+      e.message += `\n${missingReferenceErrorMessage}`;
+      this._onError(e);
+      return utils.createErrorNode(node, e);
+    }
+  }
+
+
+  /**
+   * Extracts the content's page variables, then renders it.
+   * @param content to render
+   * @param contentFilePath of the content to render
+   * @param additionalVariables if any
+   */
+  renderOuterVariables(content, contentFilePath, additionalVariables = {}) {
```

Comment:
new combined rendering methods ( `renderInner/OuterVariables` ) - aim to get this down to 1 in part 2